lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +319,318 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
-                SubtaskFailed = true;
-            } else {
-              // Since we don't have a TaskCrashedCallback, treat a crashing
-              // subtask as having failed.
-              SubtaskFailed = true;
-            }
-          }
-
-          ExecutingTasks.erase(Pid);
-          FinishedFds.push_back(fd.fd);
-        }
-      } else if (fd.revents & POLLNVAL) {
-        // We passed an invalid fd; this should never happen,
-        // since we always mark fds as finished after calling
-        // Task::finishExecution() (which closes the Task's fd).
-        llvm_unreachable("Asked poll() to watch a closed fd");
-      }
-
-      fd.revents = 0;
-    }
+    Optional<std::vector<int>> FinishedFds =
+        readFromReadyFdsReturningFinishedOnes();
+    if (!FinishedFds)
+      return true;
 
-    // Remove any fds which we've closed from PollFds.
-    for (int fd : FinishedFds) {
-      auto predicate = [&fd] (struct pollfd &i) {
-        return i.fd == fd;
-      };
+    stopPolling(*FinishedFds);
+  }
+  return false;
+}
 
-      auto iter = std::find_if(PollFds.begin(), PollFds.end(), predicate);
-      assert(iter != PollFds.end() && "The finished fd must be in PollFds!");
-      PollFds.erase(iter);
-    }
+bool TaskShepherd::startUpSomeTasks() {
+  while (!TasksToBeExecuted.empty() &&
+         TasksBeingExecuted.size() < MaxNumberOfParallelTasks) {
+    std::unique_ptr<Task> T(TasksToBeExecuted.front().release());
+    TasksToBeExecuted.pop();
+    if (beginExecutingATask(T.get()))
+      return true;
+    startPollingFdsOfTask(T.get());
+    TasksBeingExecuted.add(std::move(T));
+  }
+  return false;
+}
+
+void TaskShepherd::startPollingFdsOfTask(const Task *T) {
+  FdsBeingPolled.push_back({T->getPipe(), POLLIN | POLLPRI | POLLHUP, 0});
+  // We should also poll T->getErrorPipe(), but this introduces timing
+  // issues with shutting down the task after reading getPipe().
+}
+
+TaskShepherd::PollResult TaskShepherd::pollTheFds() {
+  assert(!FdsBeingPolled.empty() &&
+         "We should only call poll() if we have fds to watch!");
+  int ReadyFdCount = poll(FdsBeingPolled.data(), FdsBeingPolled.size(), -1);
+  callBacks.PolledAnFd();
+  if (ReadyFdCount != -1)
+    return PollResult::NoError;
+  return errno == EAGAIN || errno == EINTR ? PollResult::SoftError
+                                           : PollResult::HardError;
+}
+
+bool TaskShepherd::beginExecutingATask(Task *T) {
+  if (T->execute())
+    return true;
+  callBacks.TaskBegan(T->getPid(), T->getContext());
+  return false;
+}
+
+Optional<std::vector<int>>
+TaskShepherd::readFromReadyFdsReturningFinishedOnes() {
+  std::vector<int> finishedFds;
+  for (struct pollfd &fd : FdsBeingPolled) {
+    const int fileDes = fd.fd;
+    const short receivedEvents = fd.revents;
+    fd.revents = 0;
+    Task &T = TasksBeingExecuted.findTaskForFd(fileDes);
+    const bool fdHungUp = reactToFdEvents(receivedEvents, fileDes, T);
```

Comment:
`if (reactToFdEvents(...)) continue` seems like it'd be ok here.

Reply:
Er, I suppose `if (!reactToFdEvents(...))`. I dunno, I guess I'm illustrating the very reason why you'd name the boolean here, except I read the condition backwards (bools here are usually used to indicate errors).  Like I read this a few times and thought you had the condition inverted and were dropping hangups on the floor.

Reply:
Ok, new idea: split `reactToFdEvents` in two:

  - `void readDataIfAvailable(events, fd, task)` which does the reading part (and in the future: propagates errors from `readFromPipes`, unlike presently which drops them on the floor.
  - `bool didTaskHangup(events)` which just evaluates whether to complete the task.

Reply:
I like the idea. I have implemented it, at the cost of some event-parsing duplication, but I think it's still better.